js/src/name/publish.js

```diff
@@ -0,0 +1,103 @@
+/* eslint-env mocha */
+'use strict'
+
+const hat = require('hat')
+
+const { fixture } = require('./utils')
+const { spawnNodeWithId } = require('../utils/spawn')
+const { getDescribe, getIt, expect } = require('../utils/mocha')
+
+module.exports = (createCommon, options) => {
+  const describe = getDescribe(options)
+  const it = getIt(options)
+  const common = createCommon()
+
+  describe('.name.publish', function () {
+    const keyName = hat()
+    let ipfs
+    let nodeId
+
+    before(function (done) {
+      // CI takes longer to instantiate the daemon, so we need to increase the
+      // timeout for the before step
+      this.timeout(60 * 1000)
+
+      common.setup((err, factory) => {
+        expect(err).to.not.exist()
+
+        spawnNodeWithId(factory, (err, node) => {
+          expect(err).to.not.exist()
+
+          ipfs = node
+          nodeId = node.peerId.id
+
+          ipfs.files.add(fixture.data, { pin: false }, done)
+        })
+      })
+    })
+
+    after((done) => common.teardown(done))
+
+    it('should publish an IPNS record with the default params', (done) => {
+      this.timeout(50 * 1000)
+
+      const value = fixture.cid
+
+      ipfs.name.publish(value, (err, res) => {
+        expect(err).to.not.exist()
+        expect(res).to.exist()
+        expect(res.name).to.equal(nodeId)
+        expect(res.value).to.equal(`/ipfs/${value}`)
+
+        done()
+      })
+    })
+
+    it('should publish correctly when the file was not added but resolve is disabled', (done) => {
+      this.timeout(50 * 1000)
+
+      const value = 'QmPFVLPmp9zv5Z5KUqLhe2EivAGccQW2r7M7jhVJGLZoZU'
+
+      const options = {
+        resolve: false,
+        lifetime: '1m',
+        ttl: '10s',
+        key: 'self'
+      }
+
+      ipfs.name.publish(value, options, (err, res) => {
+        expect(err).to.not.exist()
+        expect(res).to.exist()
+        expect(res.name).to.equal(nodeId)
+        expect(res.value).to.equal(`/ipfs/${value}`)
+
+        done()
+      })
+    })
+
+    it('should recursively resolve to an IPFS hash', (done) => {
```

Comment:
This test might need to be renamed or rewritten. Was the intention to ensure that publishing `/ipfs/rootHash/path/to/file` will actually result in `/ipfs/fileHash` being published as the value? There doesn't seem to be anything recursive going on here.

Reply:
**Recursive:** Resolve until the result is not an IPNS name. Default: “false”. Required: no. (according to the docs at https://ipfs.io/docs/api/#apiv0namepublish)

This test aims to do the following:

1) Publish an `ipfs path` with the ID of the peer as the key
Result: `/ipns/{PeerID}` => `/ipfs/{CID}`

2) Publish the previously `ipns path` with the ID of a new key created
Result: `/ipns/{KeyID}` => `/ipns/{PeerID}`

3) Resolve the last recorder IPNS entry recursively:
Result: `/ipfs/{CID}`

In this test, the resolve happens recursively until the stop condition (obtaining an `ipfs path` or reaching a maximum number of tries constant). This way, the first resolve of `/ipns/{KeyID}` will result in `/ipns/{PeerID}`, which is not an `ipfs path`. Accordingly, the resolve continues, and resolving `/ipns/{PeerID}` results in `/ipfs/{CID}`. As we got an `ipfs path`, this is returned.

Maybe the test description is not the most understandable, but for the people who know what the `recursive` parameter aims to, will understand the logic. If you have any suggestion for renaming it, I will happily change it 🙂

Reply:
Wrong test name in fact, this was for `name/resolve`